mmdet3d/datasets/custom_3d_seg.py

```diff
@@ -354,22 +357,46 @@ def evaluate(self,
         assert isinstance(
             results[0], dict
         ), f'Expect elements in results to be dict, got {type(results[0])}.'
+
+        # we need to load mask annotation so set test_mode as False
+        original_test_mode = self.test_mode
+        self.test_mode = False
+        if pipeline is not None:
+            pipeline = Compose(pipeline)
+            original_pipeline = self.pipeline if \
+                hasattr(self, 'pipeline') else None  # save the original one
+            self.pipeline = pipeline  # set new pipeline for data loading
```

Comment:
Must we change the original self.pipeline and self.test_mode?

Reply:
Good question. Usually we call ```evaluate()``` and ```show()``` in eval or test, where ```self.pipeline``` is often ```test_pipeline``` (in ```config.py```). As you know, ```test_pipeline``` often involves pre-processing transforms of points/img (e.g. ```PointRangeFilter```, ```Normalize```) and doesn't have ```LoadAnnotation``` functions. Therefore, we need to change the original ```self.pipeline```.

As for ```self.test_mode```, it's ```True``` in eval and test, so dataset won't load annotation information (see [here](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/datasets/custom_3d_seg.py#L123-L125)). Therefore, we need to set it as ```False``` to load the path of GT seg mask.

Essentially, the difference stems from the GT in det and seg task. In det, gt bbox is stored in ```self.data_infos``` (see [here](https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/datasets/scannet_dataset.py#L83-L86)), so we don't need ```LoadAnnotation3D``` function in ```eval_pipeline```. But GT seg mask is not stored in ```self.data_infos``` and could only be loaded from file, so we need to set the pipeline and test_mode.